docs/spec/escrow.rst

```diff
@@ -0,0 +1,312 @@
+
+The Escrow module allows users to lockup tokens in exchange for a non-fungible voting asset. The total "power" of this asset decays linearly as the lock approaches expiry - calculated based on the block height. Historic points for the linear function are recorded each time a user's balance is adjusted which allows us to re-construct voting power at a particular point in time.
+
+This architecture was adopted from Curve, see: `Vote-Escrowed CRV (veCRV) <https://curve.readthedocs.io/dao-vecrv.html>`_.
+
+Step-by-step
+------------
+
+1. A user may lock any amount of some currency up to a maximum lock period.
+2. Both the amount and the unlock time may be increased to improve voting power.
+3. The user may unlock their fungible asset after the lock has expired.
+
+Data Model
+~~~~~~~~~~
+
+Constants
+---------
+
+.. _span:
+
+Span
+....
+
+Used to round heights.
+
+.. _maxPeriod:
+
+MaxPeriod
+.........
+
+The maximum period for lockup.
+
+Scalars
+-------
+
+.. _escrow-scalar-epoch:
+
+Epoch
+.....
+
+The current global epoch for ``PointHistory``.
+
+Maps
+----
+
+.. _escrow-map-locked:
+
+Locked
+......
+
+Stores the ``amount`` and ``end`` block for the account's lock.
+
+.. _escrow-map-point-history:
+
+PointHistory
+............
+
+Stores the global ``bias``, ``slope`` and ``ts`` at a particular point in history,
+
+.. _escrow-map-user-point-history:
+
+UserPointHistory
+................
+
+Stores the ``bias``, ``slope`` and ``ts`` for an account at a particular point in history,
+
+.. _escrow-map-user-point-epoch:
+
+UserPointEpoch
+..............
+
+Stores the current epoch for an account.
+
+.. _escrow-map-slope-changes:
+
+SlopeChanges
+............
+
+Stores scheduled changes of slopes for ending locks.
+
+Structs
+-------
+
+LockedBalance
+.............
+
+The ``amount`` and ``end`` height for a locked balance.
+
+.. tabularcolumns:: |l|L|
+
+==============  ============  ========================================================	
+Parameter       Type          Description                                            
+==============  ============  ========================================================
+``amount``      Balance       The amount deposited to receive vote-escrowed tokens.
+``end``         BlockNumber   The end height after which the balance can be unlocked.
+==============  ============  ========================================================
+
+Point
+.....
+
+The ``bias``, ``slope`` and ``ts`` for our linear function.
+
+.. tabularcolumns:: |l|L|
+
+==============  ============  ========================================================	
+Parameter       Type          Description                                            
+==============  ============  ========================================================
+``bias``        Balance       The bias for the linear function.
+``slope``       Balance       The slope for the linear function.
+``ts``          BlockNumber   The current block height when this point was stored.
```

Comment:
Nitpick: maybe name this `timestamp` for better readability.

Reply:
+1